File: 5-Problem-A-Day/39. Combination Sum.java
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> ans = new LinkedList<>();
        backtrack(candidates, 0, 0, target, new LinkedList<Integer>(), ans);
        return ans;
    }

    private void backtrack(int[] candidates, int start, int sum, int target, List<Integer> curr,
            List<List<Integer>> ans) {
        if (sum == target)
            ans.add(new LinkedList<Integer>(curr));
        else {
            for (int i = start; i < candidates.length; i++) {
                if (sum + candidates[i] <= target) {
                    curr.add(candidates[i]);
                    backtrack(candidates, i, sum + candidates[i], target, curr, ans);
                    curr.remove(curr.size() - 1);
                }
            }
        }
    }
}

File: 2024/39. Combination Sum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<int> curr;
        vector<vector<int>> res;
        backtrack(candidates, 0, target, curr, res);
        return res;
    }

    void backtrack(vector<int>& candidates, int start, int target,
                   vector<int>& curr, vector<vector<int>>& res) {
        if (target == 0) {
            res.push_back(curr);
            return;
        }
        if (target < 0) {
            return;
        }
        for (int i = start; i < candidates.size(); i++) {
            curr.push_back(candidates[i]);
            backtrack(candidates, i, target - candidates[i], curr, res);
            curr.pop_back();
        }
    }
};